enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,613 @@
+
+When MicroShift is up and running healthy, it will persist its own version and ID of current ostree commit into data dir as a file, e.g.:
+```json
+{
+  "microshift": "4.14.0",
+  "ostree_commit": "1a2b",
+}
+```
+
+MicroShift version will be used on consequent reboots to determine:
+- is upgrade is allowed or blocked
+- should storage migration be performed
+
+ID of ostree commit will be used to... TODO: Think if we can use it for "System rollback and failed data restore" scenario
+
+### Allowing and blocking upgrades (change of MicroShift version)
+
+Decision to allow or block an upgrade (change) will be based on persisted MicroShift version.
+- If binary's version is newer, allow storage migration.
+- If binary's version is older, refuse to start MicroShift.
+- If persisted version is on a list (embedded in binary) of blocked version migrations, refuse to start MicroShift.
+- If persisted version is on a list (embedded in binary) of allowed version migrations, allow storage migration.
+- If persisted version is missing and version of binary is 4.14, assume 4.13 was persisted and allow storage migration.
+
+### Data migration
+
+Data migration shall include:
+- Storage version - update of Kubernetes objects by reading Resource in older version and writing newer version
+  - Leveraging existing [Kube Storage Version Migrator](https://github.com/openshift/kubernetes-kube-storage-version-migrator) and [its Operator](https://github.com/openshift/cluster-kube-storage-version-migrator-operator).
+- etcd schema - comparing v2 and v3, most likely if/when etcd v4 is released
+  - etcd project documents how to migrate from v2 to v3, but we'll also ask OpenShift etcd team for guidance.
+
+### Open Questions [optional]
+
+- Do we want to persist ostree commit ref in the metadata?
+  - Can we make some part of the process safer, more robust, by comparing previous and current commit ref?
+  - Are we good with having previous boot MicroShift version persisted?
+- If green/red info is not persisted, and it's not a first boot (/var/lib/microshift exists), what should we do?
+  - Compare versions (metadata vs binary)
+    - Same - start cluster
+    - Different - refuse to start? Try to migrate (but create backup before)?
+
+- Should we use greenboot's green/red scripts to persist action for next boot?
+  - They have no information about what failed, so if it wasn't MicroShift, then we do unnecessary restore (possibly losing some data)
+
+- Do we want to persist the history of boots or defer to checking `boot_counter`?
+  - Red + no `boot_counter` => first boot after rollback
+  - Do we want to read kernel cmdline?
+    - Seems to be internal detail of implementation
+
+- How to we avoid endless reboots in any case? Are they possible?
+  - Maybe we can persist short history of previous boots/actions on boot?
+    - How that would protect us? If we refuse the start, greenboot checks will fail, so maybe we need override it?
```

Comment:
I mention this above, but this feels like something Greenboot should be managing. Does it? How does that work?

Reply:
Please see https://github.com/openshift/enhancements/pull/1387#discussion_r1174330557